data-runtime/src/main/java/io/micronaut/data/runtime/operations/internal/sql/AbstractSqlRepositoryOperations.java

```diff
@@ -214,24 +217,44 @@ protected <T, R> PS prepareStatement(StatementSupplier<PS> statementFunction,
      * @param index             The index
      * @param dataType          The data type
      * @param value             The value
-     * @param dialect           The dialect
+     * @param storedQuery       The SQL stored query
      */
-    protected void setStatementParameter(PS preparedStatement, int index, DataType dataType, Object value, Dialect dialect) {
+    protected void setStatementParameter(PS preparedStatement, int index, DataType dataType, Object value, SqlStoredQuery<?, ?> storedQuery) {
+        Dialect dialect = storedQuery.getDialect();
         switch (dataType) {
             case UUID:
                 if (value != null && dialect.requiresStringUUID(dataType)) {
                     value = value.toString();
                 }
                 break;
             case JSON:
-                if (value != null && !value.getClass().equals(String.class)) {
-                    if (jsonMapper == null) {
-                        throw new IllegalStateException("For JSON data types support Micronaut JsonMapper needs to be available on the classpath.");
+                if (value != null) {
+                    boolean handled = false;
+                    // In case parameter is JsonDataObject telling we need to find custom SqlJsonValueMapper
```

Comment:
There may be probably better way but since this can be object in the method or parameter inside an object, then didn't have better idea than use marker interface that will tell us how to serialize it to JSON.

Reply:
This doesn't look good. We should be able to serialize simple `Map`. Why is it needed?

Reply:
Because if we have binary (or other custom) mapper on the classpath it will be applied to all JSON serialization/deserialization and it would be different from current behavior which does from/to string serde. I thought it would be safe to mark object we want to map.
For one, JSON tests were failing because these were writing Map and other objects to String/CLOB.

We could change Oracle table schema creation to create JSON instead of CLOB and then it could be acceptable.

Reply:
Yes, then we might need to have JSON_BINARY data type or something like that? Because for all these cases data type is JSON and we don't have enough information how to pick mapper.

Reply:
What did you have in mind, how we could configure that? Considering it should work fast enough etc. What are opt in variants you have in mind, don't know what we could expect in future. I went (as usual) for simplest option to support object read/write in JSON.

Reply:
I have changed this a bit, added an annotation (not the best name maybe) that will mark we need to check for specific JSON value mapping when setting parameter during save.